content/en/docs/components/pipelines/author-a-pipeline/components.md

```diff
@@ -0,0 +1,225 @@
+
+2. Include type annotations for the function parameters and return values.
+
+    Type annotations indicate what the component inputs and outputs are and tells the KFP lightweight component executor how to serialize and deserialize the data as it is passed within a pipeline. This also (optionally) allows the KFP DSL compiler to type check your pipeline.
+
+    Valid parameter annotations include `str`, `int`, `float`, `bool`, `dict`, `list`, `OutputPath`, `InputPath`, `Input[<Artifact>]`, and `Output[<Artifact>]`.
+    
+    Valid return annotations include `str`, `int`, `float`, `bool`, `dict`, and `list`. You may also specify multiple return values by using these annotations within a `typing.NamedTuple`.
+
+    For detailed discussion on type annotations and runtime behavior, see [Data Passing][data-passing].
+
+3. Decorate your function with the `@kfp.dsl.component` decorator.
+    This decorator transforms a Python function into a component that can be used within a pipeline.
+
+    For a comprehensive list of `@kfp.dsl.component` decorator arguments, see the DSL [reference documentation][dsl-reference-documentation].
+
+
+The following is an example of a lightweight component that trains a model on an existing input `Dataset` artifact for `num_epochs` epochs, then saves the output `Model` artifact.
+
+```python
+from kfp import dsl
+
+@dsl.component(
+    base_image='python:3.7',
+    packages_to_install=['tensorflow==2.9.1']
+)
+def train_model(
+    dataset: Input[Dataset],
+    model: Output[Model],
+    num_epochs: int,
+):
+    from tensorflow import keras
+    
+    # load and process the Dataset artifact
+    with open(dataset.path) as f:
+        x, y = ...
+
+    my_model = keras.Sequential(
+        [
+            layers.Dense(4, activation='relu', name='layer1'),
+            layers.Dense(3, activation='relu', name='layer2'),
+            layers.Dense(2, activation='relu', name='layer2'),
+            layers.Dense(1, name='layer3'),
+        ]
+    )
+
+    my_model.compile(...)
+    # train for num_epochs
+    my_model.fit(x, y, epochs=num_epochs)
+    
+    # save the Model artifact
+    my_model.save(model.path)
+```
+
+Notice the `base_image` argument to the `@kfp.dsl.component` decorator. Despite not having the word "container" in its name, lightweight components are still executed as a container at runtime. The `@kfp.dsl.component` decorator mereley provides a convient Pythonic interface to defining this container image, command, and arguments. [`python:3.7`][python-docker-image] is the default image, but can be changed to any image accessible to the executing backend, as long as the image has a Python interpreter available as `python3`. Packages in `packages_to_install` will be pip installed at container runtime.
+
+**When to use?** Lightweight components should be used any time your component implementation can be written as a standalone Python function and does not require an abundance of source code.
```

Comment:
Lightweight component is preferred for quick demo, notebook use. 
For production usage, more importantly components to be shared and reused, we recommend containerize the implementation even if it'w written by Python.

https://docs.google.com/presentation/d/1uHArEC7eMgZMUdy_V-h2mHTiU1N5Z5JA48-OEbkGs3g/edit?resourcekey=0-HrhC7BsFkEGurXi0c02fOA#slide=id.gb7fca233c8_0_73p

Reply:
> can be written as a standalone Python function and does not require an abundance of source code.

Does this capture the underlying rationale for why lightweight components are preferred for quick demo / notebook use? Or is there another reason? My sense is that if there is very minimal source code and the component function is simple that there isn't an advantage to containerizing it (and may actually be a disadvantage b/c it (a) adds a step to the component creation process and (2) adds another artifact required to use the component (private image + component YAML v just component YAML). Is there an advantage I'm not thinking of?

Either way, the suggested usage is helpful and I will add to this file.